src/include/storage/buffer_manager/bm_file_handle.h

```diff
@@ -114,8 +151,10 @@ class BMFileHandle : public FileHandle {
     std::vector<std::unique_ptr<PageState>> pageStates;
     // Each file page group corresponds to a frame group in the VMRegion.
     std::vector<common::page_group_idx_t> frameGroupIdxes;
-    std::vector<std::vector<common::page_idx_t>> pageVersions;
-    std::vector<std::unique_ptr<std::atomic_flag>> pageGroupLocks;
+    // For each page group, if it has any WAL page version, we keep a `WALPageIdxGroup` in this map.
+    // `WALPageIdxGroup` records the WAL page idx for each page in the page group.
+    // Accesses to this map is synchronized by `fhSharedMutex`.
+    std::unordered_map<common::page_group_idx_t, std::unique_ptr<WALPageIdxGroup>> walPageIdxGroups;
```

Comment:
Not sure if this change is necessary. 
```
    std::vector<std::vector<common::page_idx_t>> pageVersions;
    std::vector<std::unique_ptr<std::atomic_flag>> pageGroupLocks;
```
These data structures are a very tiny (1/1024th only even less) fraction of the total memory footprint compared to all PageSrates, so we don't need to optimize this. Your call but overall make sure you think about these ballpark numbers before you optimize such details in code.

Reply:
I agree the memory footprint is small, which is not worth optimizing. But I think using map here simplifies the logic, thus we don't need to grow the vector of WALPageIdxGroup each time we add a new page group, even without any versioned wal pages.